include/swift/Frontend/FrontendInputsAndOutputs.h

```diff
@@ -125,6 +161,40 @@ class FrontendInputsAndOutputs {
   void addInputFile(StringRef file, llvm::MemoryBuffer *buffer = nullptr);
   void addPrimaryInputFile(StringRef file,
                            llvm::MemoryBuffer *buffer = nullptr);
+
+  // Outputs
+
+private:
+  friend class ArgsToFrontendOptionsConverter;
+
+  void setMainOutputs(ArrayRef<std::string> outputFiles);
+
+public:
+  unsigned countOfInputsProducingOutput() const;
+
+  const InputFile &firstInputProducingOutput() const;
+  const InputFile &lastInputProducingOutput() const;
+
+  void forEachInputProducingOutput(
+      llvm::function_ref<void(const InputFile &)> fn) const;
+
+  std::vector<std::string> copyOutputFilenames() const;
+
+  void
+  forEachOutputFilename(llvm::function_ref<void(const std::string)> fn) const;
```

Comment:
Yes, meant to use `&`, and have added that.

Changed the name to `forEachMainOutputFilename`. There is a dilemma here: existing practice in the code uses (for instance) `outputFIlename` to mean the main output. I have tended to follow that, but then as you note, some confusion arises. Sounds like you favor using the *main* for new things and not disturbing the old. Is that right?

Reply:
Ah, yeah, that's where I'd go for now. Then we can change over existing things to specify "main" explicitly later. (Or \*sigh\* pick another word because maybe people will think the "main" output is the one that has a program entry point.)